File: sandbox_rebuild/pipeline/controllers/room_assignment.py
"""RoomAssignment — ONLY controller allowed to change authoritative room.

Rules (PIPELINE_DESIGN_CURSOR.md §3):
  - Conflict → block (return CONFLICT). Swap is explicit mode.
  - Move is atomic (transaction).
  - Every move logged to v2_room_moves forever.
  - No Displace in v1.
  - Mirrors roomCode onto v2_guests + v2_reservations when linked.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import Any
from uuid import uuid4

from ..dates import dates_overlap
from .base import BaseController, ControllerResult


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class RoomAssignmentController(BaseController):
    name = "RoomAssignment"

    def create_assignment(
        self,
        *,
        reservation_id: str,
        reservation_number: str,
        room_code: str,
        checkin: str,
        checkout: str,
        guest_id: str | None = None,
        actor: str = "sandbox",
        manual_room: bool = False,
    ) -> ControllerResult:
        """Seed helper for tests / ReservationSync handoff. Not a silent move."""
        assignment_id = f"asg_{uuid4().hex[:12]}"
        doc = {
            "assignmentId": assignment_id,
            "reservationId": reservation_id,
            "reservationNumber": reservation_number,
            "guestId": guest_id,
            "roomCode": room_code,
            "checkin": checkin,
            "checkout": checkout,
            "status": "active",
            "lockSource": "admin" if manual_room else "minihotel",
            "manualRoom": manual_room,
            "updatedAt": _now(),
            "updatedBy": actor,
            "version": 1,
        }
        conflict = self._find_conflicts(room_code, checkin, checkout, exclude_id=None)
        if conflict:
            self.log.write(
                controller=self.name,
                action="create_assignment",
                status="error",
                message="Cannot create assignment — room conflict",
                input_data={"roomCode": room_code, "checkin": checkin, "checkout": checkout},
                output_data={"conflict": conflict},
                alert=True,
            )
            return ControllerResult(False, "CONFLICT", "Room already occupied for these dates", {"conflict": conflict})

        self.store.set(self.cfg.assignments, assignment_id, doc)
        self._mirror(doc)
        self.log.write(
            controller=self.name,
            action="create_assignment",
            status="ok",
            message=f"Created assignment {assignment_id} in {room_code}",
            input_data={"reservationId": reservation_id, "roomCode": room_code},
            output_data={"assignmentId": assignment_id},
        )
        return ControllerResult(True, "CREATED", "Assignment created", {"assignment": doc})

    def move(
        self,
        *,
        assignment_id: str,
        to_room: str,
        actor: str = "sandbox",
        expected_version: int | None = None,
    ) -> ControllerResult:
        return self._mutate(
            mode="move",
            assignment_id=assignment_id,
            to_room=to_room,
            actor=actor,
            expected_version=expected_version,
            other_assignment_id=None,
        )

    def swap(
        self,
        *,
        assignment_id: str,
        other_assignment_id: str,
        actor: str = "sandbox",
    ) -> ControllerResult:
        return self._mutate(
            mode="swap",
            assignment_id=assignment_id,
            to_room=None,
            actor=actor,
            expected_version=None,
            other_assignment_id=other_assignment_id,
        )

    def release_to_minihotel(self, *, assignment_id: str, actor: str = "sandbox") -> ControllerResult:
        doc = self.store.get(self.cfg.assignments, assignment_id)
        if not doc:
            return self._fail("release_to_minihotel", "NOT_FOUND", "Assignment not found", {"assignmentId": assignment_id})
        doc["manualRoom"] = False
        doc["lockSource"] = "minihotel"
        doc["updatedAt"] = _now()
        doc["updatedBy"] = actor
        doc["version"] = int(doc.get("version") or 1) + 1
        self.store.set(self.cfg.assignments, assignment_id, doc)
        self._mirror(doc)
        self._write_move_audit(
            mode="release_to_minihotel",
            assignment=doc,
            from_room=doc["roomCode"],
            to_room=doc["roomCode"],
            actor=actor,
            before_version=int(doc["version"]) - 1,
            after_version=int(doc["version"]),
            other=None,
        )
        self.log.write(
            controller=self.name,
            action="release_to_minihotel",
            status="ok",
            message=f"Cleared manualRoom on {assignment_id}",
            input_data={"assignmentId": assignment_id},
            output_data={"manualRoom": False},
        )
        return ControllerResult(True, "RELEASED", "Now follows MiniHotel room updates", {"assignment": doc})

    # ── internals ──────────────────────────────────────────────

    def _mutate(
        self,
        *,
        mode: str,
        assignment_id: str,
        to_room: str | None,
        actor: str,
        expected_version: int | None,
        other_assignment_id: str | None,
    ) -> ControllerResult:
        inp = {
            "mode": mode,
            "assignmentId": assignment_id,
            "toRoom": to_room,
            "otherAssignmentId": other_assignment_id,
            "actor": actor,
        }

        def txn(store):
            a = store.get(self.cfg.assignments, assignment_id)
            if not a:
                raise _Abort("NOT_FOUND", "Assignment not found")
            if a.get("status") != "active":
                raise _Abort("INACTIVE", f"Assignment status is {a.get('status')}")
            if expected_version is not None and int(a.get("version") or 0) != expected_version:
                raise _Abort("VERSION_CONFLICT", "Assignment changed since you loaded it — refresh and retry")

            if mode == "swap":
                if not other_assignment_id:
                    raise _Abort("BAD_REQUEST", "swap requires other_assignment_id")
                b = store.get(self.cfg.assignments, other_assignment_id)
                if not b:
                    raise _Abort("NOT_FOUND", "Other assignment not found")
                if b.get("status") != "active":
                    raise _Abort("INACTIVE", "Other assignment is not active")
                # Swap only if date windows overlap enough that they are competing — always allowed as explicit admin choice
                room_a, room_b = a["roomCode"], b["roomCode"]
                before_a, before_b = int(a.get("version") or 1), int(b.get("version") or 1)
                a["roomCode"], b["roomCode"] = room_b, room_a
                for doc, before in ((a, before_a), (b, before_b)):
                    doc["manualRoom"] = True
                    doc["lockSource"] = "admin"
                    doc["updatedAt"] = _now()
                    doc["updatedBy"] = actor
                    doc["version"] = before + 1
                store.set(self.cfg.assignments, assignment_id, a)
                store.set(self.cfg.assignments, other_assignment_id, b)
                return {
                    "a": a,
                    "b": b,
                    "fromRoom": room_a,
                    "toRoom": room_b,
                    "beforeVersion": before_a,
                    "afterVersion": before_a + 1,
                }

            # move
            assert to_room
            if to_room == a.get("roomCode"):
                raise _Abort("NOOP", "Already in that room")
            conflicts = self._find_conflicts(
                to_room, a["checkin"], a["checkout"], exclude_id=assignment_id, store=store
            )
            if conflicts:
                raise _Abort(
                    "CONFLICT",
                    "Target room already has an overlapping stay — cancel or use swap",
                    {"conflict": conflicts[0], "conflicts": conflicts},
                )
            from_room = a["roomCode"]
            before = int(a.get("version") or 1)
            a["roomCode"] = to_room
            a["manualRoom"] = True
            a["lockSource"] = "admin"
            a["updatedAt"] = _now()
            a["updatedBy"] = actor
            a["version"] = before + 1
            store.set(self.cfg.assignments, assignment_id, a)
            return {
                "a": a,
                "b": None,
                "fromRoom": from_room,
                "toRoom": to_room,
                "beforeVersion": before,
                "afterVersion": before + 1,
            }

        try:
            result = self.store.run_transaction(txn)
        except _Abort as e:
            self.log.write(
                controller=self.name,
                action=mode,
                status="error" if e.code != "NOOP" else "warn",
                message=e.message,
                input_data=inp,
                output_data=e.data,
                alert=e.code in ("CONFLICT", "VERSION_CONFLICT"),
                severity="warn" if e.code == "CONFLICT" else "error",
            )
            return ControllerResult(False, e.code, e.message, e.data)

        # mirrors + audit outside txn copy is fine for memory store (already committed)
        self._mirror(result["a"])
        if result.get("b"):
            self._mirror(result["b"])

        self._write_move_audit(
            mode=mode,
            assignment=result["a"],
            from_room=result["fromRoom"],
            to_room=result["toRoom"],
            actor=actor,
            before_version=result["beforeVersion"],
            after_version=result["afterVersion"],
            other=result.get("b"),
        )
        if result.get("b"):
            self._write_move_audit(
                mode=mode,
                assignment=result["b"],
                from_room=result["toRoom"],
                to_room=result["fromRoom"],
                actor=actor,
                before_version=int(result["b"]["version"]) - 1,
                after_version=int(result["b"]["version"]),
                other=result["a"],
            )

        self.log.write(
            controller=self.name,
            action=mode,
            status="ok",
            message=f"{mode} {assignment_id}: {result['fromRoom']} → {result['toRoom']}",
            input_data=inp,
            output_data={
                "assignmentId": assignment_id,
                "fromRoom": result["fromRoom"],
                "toRoom": result["toRoom"],
                "version": result["afterVersion"],
            },
        )
        return ControllerResult(
            True,
            "MOVED" if mode == "move" else "SWAPPED",
            f"Room {mode} succeeded",
            {"assignment": result["a"], "other": result.get("b")},
        )

    def _find_conflicts(
        self,
        room_code: str,
        checkin: str,
        checkout: str,
        exclude_id: str | None,
        store=None,
    ) -> list[dict[str, Any]]:
        store = store or self.store
        rows = store.query(self.cfg.assignments, where=[("roomCode", "==", room_code), ("status", "==", "active")])
        conflicts = []
        for doc_id, doc in rows:
            if exclude_id and doc_id == exclude_id:
                continue
            if dates_overlap(checkin, checkout, doc.get("checkin", ""), doc.get("checkout", "")):
                conflicts.append(
                    {
                        "assignmentId": doc_id,
                        "reservationNumber": doc.get("reservationNumber"),
                        "guestId": doc.get("guestId"),
                        "roomCode": doc.get("roomCode"),
                        "checkin": doc.get("checkin"),
                        "checkout": doc.get("checkout"),
                    }
                )
        return conflicts

    def _mirror(self, assignment: dict[str, Any]) -> None:
        room = assignment.get("roomCode")
        guest_id = assignment.get("guestId")
        res_id = assignment.get("reservationId")
        if guest_id:
            self.store.set(
                self.cfg.guests,
                guest_id,
                {"roomCode": room, "assignmentId": assignment.get("assignmentId"), "updatedAt": _now()},
                merge=True,
            )
        if res_id:
            self.store.set(
                self.cfg.reservations,
                res_id,
                {
                    "roomCode": room,
                    "manualRoom": bool(assignment.get("manualRoom")),
                    "assignmentId": assignment.get("assignmentId"),
                    "updatedAt": _now(),
                },
                merge=True,
            )

    def _write_move_audit(self, **kw) -> None:
        assignment = kw["assignment"]
        self.store.add(
            self.cfg.room_moves,
            {
                "assignmentId": assignment.get("assignmentId"),
                "guestId": assignment.get("guestId"),
                "reservationId": assignment.get("reservationId"),
                "fromRoom": kw["from_room"],
                "toRoom": kw["to_room"],
                "mode": kw["mode"],
                "actor": kw["actor"],
                "at": _now(),
                "beforeVersion": kw["before_version"],
                "afterVersion": kw["after_version"],
                "otherAssignmentId": (kw["other"] or {}).get("assignmentId") if kw.get("other") else None,
            },
        )

    def _fail(self, action: str, code: str, message: str, data: dict | None = None) -> ControllerResult:
        self.log.write(
            controller=self.name,
            action=action,
            status="error",
            message=message,
            input_data=data or {},
            alert=True,
        )
        return ControllerResult(False, code, message, data)


class _Abort(Exception):
    def __init__(self, code: str, message: str, data: dict | None = None):
        super().__init__(message)
        self.code = code
        self.message = message
        self.data = data or {}

File: pipeline-functions/controllers/roomAssignment.js
'use strict';
/**
 * RoomAssignment — sole writer of authoritative room moves.
 *
 * Uses existing Firestore collections (no v2_*):
 *   reservations.roomCode (+ manualRoom, allRooms, roomVersion)
 *   checkin_guests.aptId (mirror for docs linked via matchedReservationId)
 *   room_moves/{id} (audit trail)
 *
 * Trigger: AdminAction only (`move` | `swap` | `release_to_minihotel`).
 * Conflict policy (LOCKED): block + swap only — no silent overwrite, no displace.
 */

const { datesOverlap } = require('../lib/dates');

class RoomAbort extends Error {
  constructor(code, message, data = null) {
    super(message);
    this.code = code;
    this.message = message;
    this.data = data;
  }
}

function stayDates(res) {
  return {
    checkin: res.checkin || res.checkIn || '',
    checkout: res.checkout || res.checkOut || '',
  };
}

function isActiveReservation(res) {
  const status = String(res.status || '').toUpperCase();
  return status !== 'CANCELLED';
}

function reservationSummary(id, data) {
  const dates = stayDates(data);
  return {
    reservationId: id,
    reservationNumber: data.reservationNumber || data.reservation_number || null,
    roomCode: data.roomCode || data.room || null,
    checkin: dates.checkin,
    checkout: dates.checkout,
    guest: data.guest || data.guestName || null,
  };
}

function findConflicts(reservations, roomCode, checkin, checkout, excludeId) {
  const conflicts = [];
  for (const [id, data] of reservations) {
    if (excludeId && id === excludeId) continue;
    if ((data.roomCode || data.room) !== roomCode) continue;
    if (!isActiveReservation(data)) continue;
    const dates = stayDates(data);
    if (datesOverlap(checkin, checkout, dates.checkin, dates.checkout)) {
      conflicts.push(reservationSummary(id, data));
    }
  }
  return conflicts;
}

function bumpRoomVersion(res, actor, nowIso) {
  const before = Number(res.roomVersion || 0);
  return {
    roomVersion: before + 1,
    updatedAt: nowIso,
    updatedBy: actor,
    _beforeRoomVersion: before,
  };
}

/**
 * @param {object} ctx
 * @param {() => string} ctx.nowIso
 * @param {(reservationId:string) => Promise<{id:string,data:object|null}>} ctx.getReservation
 * @param {(roomCode:string) => Promise<Map<string,object>>} ctx.listReservationsInRoom
 * @param {(reservationId:string) => Promise<string[]>} ctx.listGuestIdsForReservation
 * @param {(fn:(tx:object)=>Promise<object>) => Promise<object>} ctx.runTransaction
 * @param {(entry:object) => Promise<void>} ctx.logRun
 * @param {(audit:object) => Promise<void>} ctx.writeRoomMove
 * @param {{mode:'move'|'swap'|'release_to_minihotel', assignmentId:string,
 *          toRoomCode?:string, otherAssignmentId?:string, actor:string,
 *          expectedVersion?:number|null}} params
 * @returns {Promise<{ok:boolean, errorCode?:string, message:string, data?:object}>}
 */
async function runRoomAssignment(ctx, params) {
  const mode = params.mode;
  const assignmentId = params.assignmentId;
  const actor = params.actor || 'unknown';
  const correlationId = params.correlationId || null;
  const input = {
    mode,
    assignmentId,
    toRoomCode: params.toRoomCode || null,
    otherAssignmentId: params.otherAssignmentId || null,
    actor,
    expectedVersion: params.expectedVersion ?? null,
    correlationId,
  };

  try {
    if (mode === 'release_to_minihotel') {
      return await releaseToMinihotel(ctx, assignmentId, actor, input, correlationId);
    }
    if (mode === 'swap') {
      return await swapRooms(ctx, assignmentId, params.otherAssignmentId, actor, input, correlationId);
    }
    if (mode === 'move') {
      return await moveRoom(ctx, assignmentId, params.toRoomCode, actor, params.expectedVersion, input, correlationId);
    }
    const message = `Unknown mode: ${mode}`;
    await ctx.logRun({
      controller: 'RoomAssignment',
      action: mode || 'unknown',
      status: 'error',
      message,
      input,
    });
    return { ok: false, errorCode: 'BAD_REQUEST', message };
  } catch (err) {
    const message = err.message || String(err);
    await ctx.logRun({
      controller: 'RoomAssignment',
      action: mode || 'unknown',
      status: 'error',
      message,
      input,
    });
    return { ok: false, errorCode: 'INTERNAL', message };
  }
}

async function releaseToMinihotel(ctx, reservationId, actor, input, correlationId) {
  const nowIso = ctx.nowIso();
  const current = await ctx.getReservation(reservationId);
  if (!current.data) {
    await logRoomError(ctx, 'release_to_minihotel', input, 'NOT_FOUND', 'Reservation not found', null, correlationId);
    return { ok: false, errorCode: 'NOT_FOUND', message: 'Reservation not found' };
  }
  if (!isActiveReservation(current.data)) {
    await logRoomError(ctx, 'release_to_minihotel', input, 'INACTIVE', 'Reservation is cancelled', null, correlationId);
    return { ok: false, errorCode: 'INACTIVE', message: 'Reservation is cancelled' };
  }

  const fromRoom = current.data.roomCode || current.data.room || '';
  const versionMeta = bumpRoomVersion(current.data, actor, nowIso);

  await ctx.runTransaction(async (tx) => {
    const snap = await tx.getReservation(reservationId);
    if (!snap.data) throw new RoomAbort('NOT_FOUND', 'Reservation not found');
    const guestIds = await tx.listGuestIdsForReservation(reservationId);
    tx.updateReservation(reservationId, {
      manualRoom: false,
      roomVersion: versionMeta.roomVersion,
      updatedAt: versionMeta.updatedAt,
      updatedBy: versionMeta.updatedBy,
    });
    tx.writeRoomMove({
      reservationId,
      reservationNumber: snap.data.reservationNumber || null,
      guestIds,
      fromRoom,
      toRoom: fromRoom,
      mode: 'release_to_minihotel',
      actor,
      at: nowIso,
      beforeRoomVersion: versionMeta._beforeRoomVersion,
      afterRoomVersion: versionMeta.roomVersion,
      otherReservationId: null,
    });
    tx.logRun({
      controller: 'RoomAssignment',
      action: 'release_to_minihotel',
      status: 'ok',
      message: `Cleared manualRoom on ${reservationId}`,
      input,
      output: { reservationId, manualRoom: false, roomVersion: versionMeta.roomVersion },
      correlationId,
    });
  });

  return {
    ok: true,
    errorCode: 'RELEASED',
    message: 'Now follows MiniHotel room updates',
    data: { reservationId, manualRoom: false, roomVersion: versionMeta.roomVersion },
  };
}

async function moveRoom(ctx, reservationId, toRoomCode, actor, expectedVersion, input, correlationId) {
  if (!toRoomCode) {
    await logRoomError(ctx, 'move', input, 'BAD_REQUEST', 'toRoomCode is required');
    return { ok: false, errorCode: 'BAD_REQUEST', message: 'toRoomCode is required' };
  }

  const current = await ctx.getReservation(reservationId);
  if (!current.data) {
    await logRoomError(ctx, 'move', input, 'NOT_FOUND', 'Reservation not found');
    return { ok: false, errorCode: 'NOT_FOUND', message: 'Reservation not found' };
  }
  if (!isActiveReservation(current.data)) {
    await logRoomError(ctx, 'move', input, 'INACTIVE', 'Reservation is cancelled');
    return { ok: false, errorCode: 'INACTIVE', message: 'Reservation is cancelled' };
  }

  const fromRoom = current.data.roomCode || current.data.room || '';
  if (fromRoom === toRoomCode) {
    await logRoomWarn(ctx, 'move', input, 'NOOP', 'Already in that room');
    return { ok: false, errorCode: 'NOOP', message: 'Already in that room' };
  }

  const dates = stayDates(current.data);
  const roomRows = await ctx.listReservationsInRoom(toRoomCode);
  const conflicts = findConflicts(roomRows, toRoomCode, dates.checkin, dates.checkout, reservationId);
  if (conflicts.length) {
    await logRoomWarn(ctx, 'move', input, 'CONFLICT', 'Target room already has an overlapping stay — cancel or use swap', {
      conflict: conflicts[0],
      conflicts,
    });
    return {
      ok: false,
      errorCode: 'CONFLICT',
      message: 'Target room already has an overlapping stay — cancel or use swap',
      data: { conflict: conflicts[0], conflicts },
    };
  }

  if (expectedVersion != null && Number(current.data.roomVersion || 0) !== Number(expectedVersion)) {
    await logRoomError(ctx, 'move', input, 'VERSION_CONFLICT', 'Reservation changed since you loaded it — refresh and retry');
    return {
      ok: false,
      errorCode: 'VERSION_CONFLICT',
      message: 'Reservation changed since you loaded it — refresh and retry',
    };
  }

  const nowIso = ctx.nowIso();
  const versionMeta = bumpRoomVersion(current.data, actor, nowIso);
  const guestIds = await ctx.listGuestIdsForReservation(reservationId);

  try {
    await ctx.runTransaction(async (tx) => {
      const snap = await tx.getReservation(reservationId);
      if (!snap.data) throw new RoomAbort('NOT_FOUND', 'Reservation not found');
      if (expectedVersion != null && Number(snap.data.roomVersion || 0) !== Number(expectedVersion)) {
        throw new RoomAbort('VERSION_CONFLICT', 'Reservation changed since you loaded it — refresh and retry');
      }
      const liveDates = stayDates(snap.data);
      const liveFrom = snap.data.roomCode || snap.data.room || '';
      if (liveFrom === toRoomCode) throw new RoomAbort('NOOP', 'Already in that room');

      const [liveRows, guestIds] = await Promise.all([
        tx.listReservationsInRoom(toRoomCode),
        tx.listGuestIdsForReservation(reservationId),
      ]);
      const liveConflicts = findConflicts(liveRows, toRoomCode, liveDates.checkin, liveDates.checkout, reservationId);
      if (liveConflicts.length) {
        throw new RoomAbort(
          'CONFLICT',
          'Target room already has an overlapping stay — cancel or use swap',
          { conflict: liveConflicts[0], conflicts: liveConflicts }
        );
      }

      tx.updateReservation(reservationId, {
        roomCode: toRoomCode,
        allRooms: toRoomCode,
        manualRoom: true,
        roomVersion: versionMeta.roomVersion,
        updatedAt: versionMeta.updatedAt,
        updatedBy: versionMeta.updatedBy,
      });

      for (const guestId of guestIds) {
        tx.updateGuest(guestId, { aptId: toRoomCode, updatedAt: versionMeta.updatedAt });
      }

      tx.writeRoomMove({
        reservationId,
        reservationNumber: snap.data.reservationNumber || null,
        guestIds,
        fromRoom: liveFrom,
        toRoom: toRoomCode,
        mode: 'move',
        actor,
        at: nowIso,
        beforeRoomVersion: versionMeta._beforeRoomVersion,
        afterRoomVersion: versionMeta.roomVersion,
        otherReservationId: null,
      });
      tx.logRun({
        controller: 'RoomAssignment',
        action: 'move',
        status: 'ok',
        message: `move ${reservationId}: ${liveFrom} → ${toRoomCode}`,
        input,
        output: { reservationId, fromRoom: liveFrom, toRoom: toRoomCode, roomVersion: versionMeta.roomVersion },
        correlationId,
      });
    });
  } catch (err) {
    if (err instanceof RoomAbort) {
      const isWarn = err.code === 'CONFLICT' || err.code === 'NOOP';
      await ctx.logRun({
        controller: 'RoomAssignment',
        action: 'move',
        status: isWarn ? 'warn' : 'error',
        message: err.message,
        input,
        output: err.data || { errorCode: err.code },
        correlationId,
      });
      return { ok: false, errorCode: err.code, message: err.message, data: err.data || undefined };
    }
    throw err;
  }

  return {
    ok: true,
    errorCode: 'MOVED',
    message: 'Room move succeeded',
    data: { reservationId, fromRoom, toRoom: toRoomCode, roomVersion: versionMeta.roomVersion, affectedGuestIds: guestIds },
  };
}

async function swapRooms(ctx, reservationId, otherReservationId, actor, input, correlationId) {
  if (!otherReservationId) {
    await logRoomError(ctx, 'swap', input, 'BAD_REQUEST', 'otherAssignmentId is required');
    return { ok: false, errorCode: 'BAD_REQUEST', message: 'otherAssignmentId is required' };
  }
  if (otherReservationId === reservationId) {
    await logRoomError(ctx, 'swap', input, 'BAD_REQUEST', 'Cannot swap a reservation with itself');
    return { ok: false, errorCode: 'BAD_REQUEST', message: 'Cannot swap a reservation with itself' };
  }

  const [a, b] = await Promise.all([
    ctx.getReservation(reservationId),
    ctx.getReservation(otherReservationId),
  ]);
  if (!a.data) {
    await logRoomError(ctx, 'swap', input, 'NOT_FOUND', 'Reservation not found');
    return { ok: false, errorCode: 'NOT_FOUND', message: 'Reservation not found' };
  }
  if (!b.data) {
    await logRoomError(ctx, 'swap', input, 'NOT_FOUND', 'Other reservation not found');
    return { ok: false, errorCode: 'NOT_FOUND', message: 'Other reservation not found' };
  }
  if (!isActiveReservation(a.data) || !isActiveReservation(b.data)) {
    await logRoomError(ctx, 'swap', input, 'INACTIVE', 'One or both reservations are cancelled');
    return { ok: false, errorCode: 'INACTIVE', message: 'One or both reservations are cancelled' };
  }

  const roomA = a.data.roomCode || a.data.room || '';
  const roomB = b.data.roomCode || b.data.room || '';
  if (roomA === roomB) {
    await logRoomWarn(ctx, 'swap', input, 'NOOP', 'Both reservations are already in the same room');
    return { ok: false, errorCode: 'NOOP', message: 'Both reservations are already in the same room' };
  }

  const nowIso = ctx.nowIso();
  const versionA = bumpRoomVersion(a.data, actor, nowIso);
  const versionB = bumpRoomVersion(b.data, actor, nowIso);
  const [guestIdsA, guestIdsB] = await Promise.all([
    ctx.listGuestIdsForReservation(reservationId),
    ctx.listGuestIdsForReservation(otherReservationId),
  ]);

  try {
    await ctx.runTransaction(async (tx) => {
      const [snapA, snapB, guestIdsA, guestIdsB] = await Promise.all([
        tx.getReservation(reservationId),
        tx.getReservation(otherReservationId),
        tx.listGuestIdsForReservation(reservationId),
        tx.listGuestIdsForReservation(otherReservationId),
      ]);
      if (!snapA.data) throw new RoomAbort('NOT_FOUND', 'Reservation not found');
      if (!snapB.data) throw new RoomAbort('NOT_FOUND', 'Other reservation not found');

      const liveA = snapA.data.roomCode || snapA.data.room || '';
      const liveB = snapB.data.roomCode || snapB.data.room || '';
      if (liveA === liveB) throw new RoomAbort('NOOP', 'Both reservations are already in the same room');

      tx.updateReservation(reservationId, {
        roomCode: liveB,
        allRooms: liveB,
        manualRoom: true,
        roomVersion: versionA.roomVersion,
        updatedAt: versionA.updatedAt,
        updatedBy: versionA.updatedBy,
      });
      tx.updateReservation(otherReservationId, {
        roomCode: liveA,
        allRooms: liveA,
        manualRoom: true,
        roomVersion: versionB.roomVersion,
        updatedAt: versionB.updatedAt,
        updatedBy: versionB.updatedBy,
      });

      for (const guestId of guestIdsA) {
        tx.updateGuest(guestId, { aptId: liveB, updatedAt: versionA.updatedAt });
      }
      for (const guestId of guestIdsB) {
        tx.updateGuest(guestId, { aptId: liveA, updatedAt: versionB.updatedAt });
      }

      tx.writeRoomMove({
        reservationId,
        reservationNumber: snapA.data.reservationNumber || null,
        guestIds: guestIdsA,
        fromRoom: liveA,
        toRoom: liveB,
        mode: 'swap',
        actor,
        at: nowIso,
        beforeRoomVersion: versionA._beforeRoomVersion,
        afterRoomVersion: versionA.roomVersion,
        otherReservationId,
      });
      tx.writeRoomMove({
        reservationId: otherReservationId,
        reservationNumber: snapB.data.reservationNumber || null,
        guestIds: guestIdsB,
        fromRoom: liveB,
        toRoom: liveA,
        mode: 'swap',
        actor,
        at: nowIso,
        beforeRoomVersion: versionB._beforeRoomVersion,
        afterRoomVersion: versionB.roomVersion,
        otherReservationId: reservationId,
      });
      tx.logRun({
        controller: 'RoomAssignment',
        action: 'swap',
        status: 'ok',
        message: `swap ${reservationId} ↔ ${otherReservationId}: ${liveA} ↔ ${liveB}`,
        input,
        output: {
          reservationId,
          otherReservationId,
          fromRoom: liveA,
          toRoom: liveB,
          roomVersionA: versionA.roomVersion,
          roomVersionB: versionB.roomVersion,
        },
        correlationId,
      });
    });
  } catch (err) {
    if (err instanceof RoomAbort) {
      const isWarn = err.code === 'NOOP';
      await ctx.logRun({
        controller: 'RoomAssignment',
        action: 'swap',
        status: isWarn ? 'warn' : 'error',
        message: err.message,
        input,
        output: err.data || { errorCode: err.code },
        correlationId,
      });
      return { ok: false, errorCode: err.code, message: err.message, data: err.data || undefined };
    }
    throw err;
  }

  return {
    ok: true,
    errorCode: 'SWAPPED',
    message: 'Room swap succeeded',
    data: {
      reservationId,
      otherReservationId,
      fromRoom: roomA,
      toRoom: roomB,
      roomVersionA: versionA.roomVersion,
      roomVersionB: versionB.roomVersion,
      affectedGuestIds: [...guestIdsA, ...guestIdsB],
    },
  };
}

async function logRoomError(ctx, action, input, code, message, output = null, correlationId = null) {
  await ctx.logRun({
    controller: 'RoomAssignment',
    action,
    status: code === 'CONFLICT' || code === 'NOOP' ? 'warn' : 'error',
    message,
    input,
    output: output || { errorCode: code },
    correlationId: correlationId || input.correlationId || null,
  });
}

async function logRoomWarn(ctx, action, input, code, message, output = null, correlationId = null) {
  await ctx.logRun({
    controller: 'RoomAssignment',
    action,
    status: 'warn',
    message,
    input,
    output: output || { errorCode: code },
    correlationId: correlationId || input.correlationId || null,
  });
}

function buildLiveCtx() {
  const { getFirestore, FieldValue } = require('firebase-admin/firestore');
  const { writeSystemLog, buildSystemLogDoc } = require('../lib/logging');

  const db = getFirestore();

  async function getReservation(reservationId) {
    const snap = await db.collection('reservations').doc(reservationId).get();
    return { id: reservationId, data: snap.exists ? snap.data() : null };
  }

  async function listReservationsInRoom(roomCode) {
    const snap = await db.collection('reservations').where('roomCode', '==', roomCode).get();
    const map = new Map();
    snap.forEach((doc) => map.set(doc.id, doc.data()));
    return map;
  }

  async function listGuestIdsForReservation(reservationId) {
    const snap = await db
      .collection('checkin_guests')
      .where('matchedReservationId', '==', reservationId)
      .get();
    return snap.docs.map((d) => d.id);
  }

  async function runTransaction(fn) {
    return db.runTransaction(async (txn) => {
      const tx = {
        async getReservation(reservationId) {
          const ref = db.collection('reservations').doc(reservationId);
          const snap = await txn.get(ref);
          return { id: reservationId, data: snap.exists ? snap.data() : null };
        },
        async listReservationsInRoom(roomCode) {
          const snap = await txn.get(db.collection('reservations').where('roomCode', '==', roomCode));
          const map = new Map();
          snap.forEach((doc) => map.set(doc.id, doc.data()));
          return map;
        },
        async listGuestIdsForReservation(reservationId) {
          const snap = await txn.get(
            db.collection('checkin_guests').where('matchedReservationId', '==', reservationId)
          );
          return snap.docs.map((d) => d.id);
        },
        updateReservation(reservationId, patch) {
          const ref = db.collection('reservations').doc(reservationId);
          txn.update(ref, patch);
        },
        updateGuest(guestId, patch) {
          const ref = db.collection('checkin_guests').doc(guestId);
          txn.set(ref, patch, { merge: true });
        },
        writeRoomMove(audit) {
          const ref = db.collection('room_moves').doc();
          txn.set(ref, { ...audit, timestamp: FieldValue.serverTimestamp() });
        },
        logRun(entry) {
          const ref = db.collection('system_logs').doc();
          txn.set(ref, buildSystemLogDoc(entry));
        },
      };

      try {
        return await fn(tx);
      } catch (err) {
        if (err instanceof RoomAbort) throw err;
        throw err;
      }
    });
  }

  return {
    nowIso: () => new Date().toISOString(),
    getReservation,
    listReservationsInRoom,
    listGuestIdsForReservation,
    runTransaction,
    logRun: (entry) => writeSystemLog(db, entry),
  };
}

module.exports = {
  runRoomAssignment,
  buildLiveCtx,
  RoomAbort,
  datesOverlap,
  findConflicts,
  stayDates,
  isActiveReservation,
};
